### ISSUE TITLE
ViewRef.onDestroy broken in Angular 9, but works if Ivy disabled.

### ISSUE THREAD

**Issue by grant77** [angular/angular#36213]
# 🐞 bug report

### Affected Package
    "@angular/core": "~9.0.5",

### Is this a regression?
Yes, this worked in Angular 8.  It also works in Angular 9 if you turn off Ivy.

### Description
Registering a callback with ViewRef.onDestroy will result in the following error when the component is destroyed:

ERROR TypeError: tCleanup[i].call is not a function

## 🔬 Minimal Reproduction
I cannot seem to update the packages in stackblitz.com, so here is a short snippet:

```
import { NgModule, Component, ChangeDetectorRef, ViewRef } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

/** AppComponent */
@Component({
  selector: 'app-root',
  template: `
    <h3>Click this button and look for red in console</h3>
    <button (click)="show=!show">Toggle</button>
    <br />
    <app-dummy *ngIf="show"></app-dummy>
`
})
export class AppComponent { show = true; }

/** DummyComponent */
@Component({ selector: 'app-dummy', template: `<h1>Hello World</h1>` })
export class DummyComponent {
  constructor(cdr: ChangeDetectorRef) {
    (cdr as ViewRef).onDestroy(() => alert('Good bye!'));
  }
}

/** AppModule */
@NgModule({
  declarations: [ AppComponent, DummyComponent ],
  imports: [ BrowserModule ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

## 🔥 Exception or Error
<pre><code>
ERROR TypeError: tCleanup[i].call is not a function
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
9.0.5
</code></pre>

**Comment by destus90** [comment]
@grant77 
just to know: you can create a runnable example with Angular 9 using https://ng-run.com/

**Comment by pkozlowski-opensource** [comment]
Minimal repro: https://ng-run.com/edit/NUF1lJkP3k0UYzSXPOHL

While I can definitively see the difference in the behaviour with the ivy enabled, there are 2 quick observations:
* casting `ChangeDetectorRef` to `ViewRef` is really using private APIs and those can change in major release without an warning;
* I'm not sure why one would use a specific pattern like this when it is possible to have a `OnDestroy` hook:

```typescript
@Component({ selector: 'app-dummy', template: `<h1>Hello World</h1>` })
export class DummyComponent implements OnDestroy {  

  ngOnDestroy() {
    console.log('Good bye!');
  }
}
```

Would you mind sharing _why_ you've decided to use the particular pattern of casting to `ViewRef` and subscribing to the destroy callback in this particular way?

**Comment by grant77** [comment]
I would never use that pattern in a real app.  It was just to replicate the bug.  The real code is here:

```import { ChangeDetectorRef, FactoryProvider, SkipSelf, ViewRef } from '@angular/core';
import { ControlContainer, FormGroupDirective, NgForm } from '@angular/forms';

export const controlContainerFactoryProvider: FactoryProvider =  {
  provide: ControlContainer,
  useFactory: controlContainerFactory,
  deps: [[ControlContainer, new SkipSelf()], ChangeDetectorRef]
};

function controlContainerFactory(controlContainer: ControlContainer, cdr: ChangeDetectorRef) {
  const form = controlContainer.formDirective as NgForm | FormGroupDirective | null;
  if (form && isViewRef(cdr)) {
    const subscription = form.ngSubmit.subscribe(() => cdr.markForCheck());
    cdr.onDestroy(() => subscription.unsubscribe()); // <=== IVY does not like this!!!!!
  }
  return controlContainer;
}

function isViewRef(cdr: ChangeDetectorRef): cdr is ViewRef {
  return 'onDestroy' in cdr; // <== "instanceof ViewRef" doesn't work
}
```

Our forms are way too big in our app, so we spread the FormControls out into child components.  As everyone knows, reactive forms does not like this.  You have to inject ControlContainer in the child and do some other goofy stuff to make it work.  I am trying to remove this repetitive boilerplate code out of our app with the code above.

**Comment by kara** [comment]
As @pkozlowski-opensource points out, `onDestroy` here is technically a private API. However, the fact that it doesn't work is a bug. 

It looks like when we refactored how we stored cleanup functions, we updated `storeCleanupWithContext` but not `storeCleanupFn` (the latter is used to make `ViewRef.onDestroy` work properly). To fix this, we will need to update `storeCleanupFn` so that the context is stored in `LView` and the cleanup function itself is stored in `TView`.
